ProjectDiffScreen: Better navigation, hunk reset, UX improvements, and file list display bugfixes

MaxFangX opened the pull request:
# `ProjectDiffScreen` UX Improvements, Features, and Fixes

## Demo

[Video link](https://drive.google.com/file/d/10N3P_Vww8-8Wtrbne_xVUWDNE3ww4Faw/view?usp=sharing)

In this demo you can see me pressing tab to toggle the cursor between the file list and the diff pane, using the next and prev bindings in the diff pane to move between hunks, use the next and prev bindings in the file list to move between files, and various changes which make it easy to stage a bunch of hunks quickly simply by pressing s in the diff pane. When opening the project diff screen, your curser is put over the first unstaged diff, so you can begin staging quickly; once all hunks in a file have been staged, you automatically jump to the first unstaged diff in the next file. The overall goal is to make it super fast and seamless to stage a bunch of hunks quickly.

## Commits

### feat(ProjectDiffScreen): Add Tab to toggle focus

Add configurable keymap to switch between file list and diff preview. When switching to diff, cursor jumps to first hunk.

### feat(ProjectDiffScreen): Add hunk_alignment setting

Configurable alignment when jumping to hunks: `'top'`, `'center'`, `'bottom'`. Defaults to `'center'` for better context visibility.

### feat(ProjectDiffScreen): Prefer unstaged entries

Focus on changes the user hasn't reviewed yet. Typical workflow: stage some hunks, exit to fix something, then re-enter project diff. Cursor should resume at unstaged changes, not staged ones.

### fix: Simplify header row positioning logic

Previously, `configure_row()` set `header_win_plot.row` to `row + 1`, and `HeaderElement:mount()` compensated by subtracting 1. Now the logic is straightforward: header stays at original row, content is pushed down.

### feat: Jump to next unstaged file after staging

When staging the last hunk in a file, jump to the next unstaged file rather than staying on the now-staged entry.

### feat: Jump to next staged file after unstaging

When unstaging the last hunk in a file, jump to the next staged file rather than staying on the now-unstaged entry.

### feat(ProjectDiffScreen): Add reset_hunk command

Allows discarding individual hunks from the diff pane. Default keybinding `r`, `buffer_reset` moved to `R`.

### feat(keymap): Add nowait to buffer-local keymaps

Without `nowait`, single-key bindings like `d` can feel sluggish if the user has operator-pending mappings. Vim waits for potential follow-up keys (`dd`, `dw`, etc.) before executing.

### feat(ProjectDiffScreen): Add next/previous nav

Generalized navigation keybinds (default `J`/`K`):
- In diff pane: navigate between hunks across all files
- In file list: navigate between files (skipping folders)

Both wrap around and jump to first/last hunk when crossing files.

### feat(ProjectDiffScreen): Jump to file position on quit

The diff view is a projection of the file's state. Quitting from the diff pane now jumps to the corresponding file position, keeping context stable.

### feat: Remember entry type across sessions

Store `vgit_last_entry_type` buffer variable on quit so re-opening returns to the same staged/unstaged entry for the file.

### fix: Handle empty hunks in deleted file generators

`generate_unified_deleted` and `generate_split_deleted` crashed when passed empty hunks. Add early return guards matching the regular diff generators.

### fix(ProjectDiffScreen): Stay at next hunk/file after staging

Hunk operations (stage/unstage/reset) stay at the same index position, effectively moving to the next hunk. When a file is fully staged or unstaged, jump to the next file of the same type rather than the first.

### fix(ProjectDiffScreen): Refresh gutter signs on quit

Dispatch sync event to git_buffer_store when exiting to a file, ensuring gutter signs reflect any staging changes made in the screen.

## Omitted commit

This commit might need some more work. So hard to get it right...

<details>

### fix: Content padding for tabline offset in floating windows

When `showtabline > 0`, floating windows with `relative='editor'` don't account for the tabline taking screen space, causing the first line(s) of content to be hidden. The fix prepends empty padding lines to buffer content when the tabline is visible (1 line for file list, 2 for diff pane).

**Before**

"Staged Changes" header missing, can't see line 1 in the diff view

<img width="1689" height="1538" alt="Before" src="https://github.com/user-attachments/assets/8b3fd3f8-d01e-4037-838b-758d1e5fc923" />

**After**

Can see "Staged Changes" header, as well as line 1 in the diff view

<img width="1690" height="1536" alt="After" src="https://github.com/user-attachments/assets/507e3925-a0f3-4af0-bd15-5482d0ab271f" />

</details>

MaxFangX committed:
feat(ProjectDiffScreen): Add Tab to toggle focus

Add configurable keymap to switch between file list and diff preview.
When switching to diff, cursor jumps to first hunk.

MaxFangX committed:
feat(ProjectDiffScreen): Add hunk_alignment setting

Configurable alignment when jumping to hunks: 'top', 'center', 'bottom'.
Defaults to 'center' for better context visibility.

MaxFangX committed:
feat(ProjectDiffScreen): Prefer unstaged entries

Focus on changes the user hasn't reviewed yet. Typical workflow: stage
some hunks, exit to fix something, then re-enter project diff. Cursor
should resume at unstaged changes, not staged ones.

MaxFangX committed:
fix: Simplify header row positioning logic

Previously, configure_row() set header_win_plot.row to row + 1, and
HeaderElement:mount() compensated by subtracting 1. Now the logic is
straightforward: header stays at original row, content is pushed down.

Also adds documentation files for known issues and conventions.

MaxFangX committed:
feat: Jump to next unstaged file after staging

When staging the last hunk in a file, jump to the next unstaged file
rather than staying on the now-staged entry.

MaxFangX committed:
feat(ProjectDiffScreen): Add reset_hunk command

Allows discarding individual hunks from the diff pane.
Default keybinding 'r', buffer_reset moved to 'R'.

MaxFangX committed:
feat(keymap): Add nowait to buffer-local keymaps

Without nowait, single-key bindings like 'd' can feel sluggish if
the user has operator-pending mappings. Vim waits for potential
follow-up keys (dd, dw, etc.) before executing.

MaxFangX committed:
feat(ProjectDiffScreen): Add next/previous nav

Generalized navigation keybinds (default J/K):
- In diff pane: navigate between hunks across all files
- In file list: navigate between files (skipping folders)

Both wrap around and jump to first/last hunk when crossing files.

MaxFangX committed:
feat: Jump to next staged file after unstaging

When unstaging the last hunk in a file, jump to the next staged file
rather than staying on the now-unstaged entry.

MaxFangX committed:
feat(ProjectDiffScreen): Jump to file position on quit

The diff view is a projection of the file's state. Quitting from the diff pane now jumps to the corresponding file position, keeping context stable.

MaxFangX committed:
feat: Remember entry type across sessions

Store vgit_last_entry_type buffer variable on quit so re-opening
returns to the same staged/unstaged entry for the file.

MaxFangX committed:
fix: Handle empty hunks in deleted file generators

`generate_unified_deleted` and `generate_split_deleted` crashed
when passed empty hunks. Add early return guards matching the
regular diff generators.

MaxFangX commented:
Rebased. Would you like me to split anything out into separate PRs?

tanvirtin commented:
> Rebased. Would you like me to split anything out into separate PRs?

Appreciate all your efforts, one PR should be good enough. Apologies for the delay on the review as I have been away travelling. I should have some time to sit down and review it this coming weekend!

MaxFangX committed:
fix: Stay at next hunk/file after staging

Hunk operations (stage/unstage/reset) stay at the same index position,
effectively moving to the next hunk. When a file is fully staged or
unstaged, jump to the next file of the same type rather than the first.

MaxFangX committed:
fix(ProjectDiffScreen): Stay at next hunk/file after staging

- Hunk operations now stay at same index (the next hunk)
- File operations now go to next file in order, not first file

MaxFangX commented:
Removed a commit which fixed a visual display issue when a tabline exists in the buffers 'behind' the project diff screen, as it appears to introduce other visual issues. This one is especially hard to fix, still iterating on it.

No rush at all on the review :)

MaxFangX committed:
fix(ProjectDiffScreen): Refresh gutter on quit

Dispatch sync event to git_buffer_store when exiting to a file,
ensuring gutter signs reflect any staging changes made in the screen.

MaxFangX commented:
Managed to fix the display issue. That was a tough one. 😅 Pushed the commit, and included some screenshots of before / after.

tanvirtin commented on lua/vgit/settings/project_diff_preview.lua:
```diff
@@ -26,6 +28,10 @@ return Config({
       desc = 'Unstage hunk'
     },
+    buffer_hunk_reset = {
+      key = 'r',
```
```suggestion
      key = 'gr',
```

tanvirtin commented on lua/vgit/settings/project_diff_preview.lua:
```diff
@@ -15,7 +17,7 @@ return Config({
     },
     buffer_reset = {
-      key = 'r',
+      key = 'R',
```
Let's keep this as `r` and have `buffer_hunk_reset` be `gr` to avoid mapping conflict with `reset_all` and keymap consistency for hunk commands being prefixed by `g`.

```suggestion
      key = 'r',
```

tanvirtin commented on lua/vgit/features/screens/ProjectDiffScreen/init.lua:
```diff
@@ -111,13 +126,43 @@ function ProjectDiffScreen:stage_hunk()
+        self:move_to(function(status)
+          return status.filename == filename
+        end)
+      end
```
The condition for `next_file` should cover moving to the next unstaged file (when there are NO other unstaged files and we just staged the last unstaged hunk).

Let me know otherwise, then we can also apply it to all other similar methods.

```suggestion
      self:move_to(function(status)
        return status.filename == filename
      end)
```

MaxFangX replied:
Kept the next_file logic (prefer jumping to next file of same type), simplified the else fallback as suggested, and extracted restore_hunk_position for reusability. Applied to stage_hunk, unstage_hunk, and reset_hunk.

tanvirtin commented on lua/vgit/features/screens/ProjectDiffScreen/init.lua:
```diff
@@ -111,13 +126,43 @@ function ProjectDiffScreen:stage_hunk()
+      local hunk_alignment = project_diff_preview_setting:get('hunk_alignment')
+      self.diff_view:move_to_hunk(target, hunk_alignment)
+    end
+  end
```
Has some potential to become a method for reusability.

MaxFangX replied:
Extracted `restore_hunk_position`.

tanvirtin commented on lua/vgit/ui/views/DiffView.lua:
```diff
@@ -9,6 +9,14 @@ local DiffComponent = require('vgit.ui.components.DiffComponent')
+-- Compensates for tabline taking screen space but not being accounted for
+-- in floating window positioning with relative='editor'. Needs 2 lines because
+-- the diff content is pushed down by the HeaderElement.
+function DiffView:get_tabline_padding()
```
Noticing some offsets caused by this

<img width="883" height="384" alt="Screenshot 2025-12-24 at 10 55 58 PM" src="https://github.com/user-attachments/assets/ca2804b3-50e1-4634-acb9-f6140c0f08f7" />
<img width="883" height="384" alt="Screenshot 2025-12-24 at 11 00 47 PM" src="https://github.com/user-attachments/assets/b2210d22-899f-4353-8a0f-e98bea3821fd" />

MaxFangX replied:
I removed the visual adjustment commit (`fix: Content padding for tabline offset in floating windows`). It fixed all of the worst bugs, but I guess there are always more... Personally I think having an extra space is better than having lines hidden due to being covered by the top bar, but I'd like to prioritize getting the rest merged first.

tanvirtin reviewed (commented):
Thanks all these improvements. Loving the new `J/K` keybinds.

Left some small comments otherwise looking good!

MaxFangX commented:
Implemented all comments, except for the one noting the visual display bug; for that one, I just removed the commit. I could continue iterating on it locally, unless you'd like me to open a separate PR for that and merge it anyway?
